pkg/auth/userregistry/identitymapper/mapper.go

```diff
@@ -0,0 +1,71 @@
+package identitymapper
+
+import (
+	"fmt"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+
+	authapi "github.com/openshift/origin/pkg/auth/api"
+	"github.com/openshift/origin/pkg/user"
+	userapi "github.com/openshift/origin/pkg/user/api"
+	identityregistry "github.com/openshift/origin/pkg/user/registry/identity"
+	userregistry "github.com/openshift/origin/pkg/user/registry/user"
+	mappingregistry "github.com/openshift/origin/pkg/user/registry/useridentitymapping"
+)
+
+type UserStrategyType string
+
+const (
+	// StrategyTypeLookup does not provision a new identity or user, it only allows identities already associated with users
+	StrategyTypeLookup UserStrategyType = "lookup"
+
+	// StrategyTypeClaim associates a new identity with a user with the identity's preferred username
+	// if no other identities are already associated with the user
+	StrategyTypeClaim UserStrategyType = "claim"
+
+	// StrategyTypeAdd associates a new identity with a user with the identity's preferred username,
+	// adding to any existing identities associated with the user
+	StrategyTypeAdd UserStrategyType = "add"
+
+	// StrategyTypeReplace associates a new identity with a user with the identity's preferred username,
+	// replacing any existing identities associated with the user
+	StrategyTypeReplace UserStrategyType = "replace"
+
+	// StrategyTypeGenerate finds an available username for a new identity, based on its preferred username
+	// If a user with the preferred username already exists, a unique username is generated
+	StrategyTypeGenerate UserStrategyType = "generate"
+)
+
+type UserStrategy interface {
+	UserForIdentity(ctx kapi.Context, preferredUserName string, identity *userapi.Identity) (*userapi.User, error)
+}
+
+// NewIdentityUserMapper returns an IdentityMapper that does the following:
+// 1. Returns an existing user if the identity exists and is associated with an existing user
+// 2. Returns an error if the identity exists and is not associated with a user (or is associated with a missing user)
+// 3. Handles new identities according to the requested strategy
+func NewIdentityUserMapper(identities identityregistry.Registry, users userregistry.Registry, strategy UserStrategyType) (authapi.UserIdentityMapper, error) {
+	initUser := user.NewDefaultUserInitStrategy()
```

Comment:
Doc when the `user.Initializer` is used and what the interface is for.  It's currently undocumented.